### DISCUSSION TITLE
how to create a large eBPF map with max_entries set to 200,000,000 ？

### DISCUSSION THREAD

**Discussion by Voyager-Zhang** [cilium/ebpf#1442] [Q&A] [answered]
I make an eBPF map with max_entries to 200,000,000.
My machine has enough memory, but it seems that there are system parameters limiting the use of large map sizes. Which specific parameters might these be?

Here is my error said：

load objects error, error is  field TcEnd: program tc_end: map cache_map: map create: argument list too long

**Comment by ouvaa** [comment]
from chatgpt4

The error message you provided, "load objects error, error is field TcEnd: program tc_end: map cache_map: map create: argument list too long," suggests that there might be a limitation on the size of the argument list being passed to the tc command, which is commonly used to interact with the Traffic Control subsystem in Linux.

When dealing with large eBPF maps, there are indeed several system parameters that could potentially limit their use:

Maximum Argument List Size: As indicated by the error message, there might be a limitation on the maximum size of the argument list that can be passed to system commands such as tc. This limitation is often controlled by the ARG_MAX parameter in the kernel, which determines the maximum size of the argument list for a new process. When creating or interacting with large eBPF maps, the size of the arguments passed to system commands could exceed this limit, leading to failures like the one you encountered.
Maximum Map Size: Although you mentioned that your machine has sufficient memory, there might still be constraints imposed by the kernel on the maximum size of eBPF maps. While the max_entries parameter defines the maximum number of entries in a map, there could be additional limitations on the total memory that can be allocated for all eBPF maps combined.
Resource Limits: The operating system imposes various resource limits to ensure fair resource allocation and prevent abuse. These limits, specified through parameters such as RLIMIT_MEMLOCK, RLIMIT_STACK, and RLIMIT_DATA, could affect the amount of memory that can be locked into physical memory, the size of the stack for each process, and the total amount of data memory that can be used by a process, respectively. Exceeding these limits could lead to failures when working with large eBPF maps.
To address the error you encountered, you should investigate the specific system parameters mentioned above and ensure that they are configured to accommodate the size of the eBPF maps you intend to create. Additionally, you may need to optimize your usage of system commands to work within the constraints imposed by these parameters.

**Reply by lmb** [reply]
I'm pretty sure this answer is wrong. Please don't post the output of an AI model without understanding the context and / or verifying the accuracy.

**Comment by ti-mo** [comment, marked as answer]
@Voyager-Zhang This is not a question specific to ebpf-go. To understand the kernel's limitations, you'll have to read the relevant kernel code. Closing this.
